components/tidb_query_codegen/src/rpn_function.rs

```diff
@@ -1,11 +1,125 @@
 // Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
 
+//! Implementation of the `rpn_fn` attribute macro.
+//!
+//! The macro should be applied to a function:
+//!
+//! ```rust
+//! #[rpb_fn]
+//! fn foo(x: &Option<u32>) -> Result<Option<u8>> {
+//!     Ok(None)
+//! }
+//! ```
+//!
+//! The expanded function implements an operation in the coprocessor.
+//!
+//! ## Arguments to macro
+//!
+//! If neither `varg` or `raw_varg` are supplied, then the generated arguments
+//! follow from the supplied function's arguments. Each argument must have a type
+//! `&Option<T>` for some `T`.
+//!
+//! ### `varg`
+//!
+//! The RPN operator takes a variable number of arguments. The arguments are passed
+//! as a `&[&Option<T>]`.
+//!
+//! ### `raw_varg`
+//!
+//! The RPN operator takes a variable number of arguments. The arguments are passed
+//! as a `&[ScalarValueRef]`.
+//!
+//! ### `min_args`
+//!
+//! The minimum number of arguments. The macro will generate code to check this
+//! as part of validation. Only valid if `varg` or `raw_varg` is also used.
+//! E.g., `#[rpn_fn(varg, min_args = 2)]`
+//!
+//! ### `extra_validator`
+//!
+//! A function name for custom validation code to be run when an operation is
+//! validated. The validator function should have the signature `&tipb::Expr -> Result<()>`.
+//! E.g., `#[rpn_fn(raw_varg, extra_validator = json_object_validator)]`
+//!
+//! ### `capture`
+//!
+//! An array of argument names which are passed from the caller to the expanded
+//! function. The argument names must be in scope in the generated `eval` or `run`
+//! methods. Currently, that includes the following arguments (the supplied
+//! function must accept these arguments with the corresponding types, in
+//! addition to any other arguments):
+//!
+//! * `ctx: &mut expr::EvalContext`
+//! * `output_rows: usize`
+//! * `args: &[rpn_expr::RpnStackNode<'_>]`
+//! * `extra: &mut rpn_expr::RpnFnCallExtra<'_>`
+//!
+//! E.g., `#[rpn_fn(capture = [ctx, extra])]`.
+//!
+//!
+//! ## Generated code
```

Comment:
Actually there are documentations in `components/tidb_query/src/rpn_expr/types`. Would you like to make some integrations like remove some duplicate contents or give some links?

Reply:
I plan to cover the other coprocessor crates in follow-up PRs and will include some links, etc. then. Is that OK?

Reply:
@nrc I mean, comments in `components/tidb_query/src/rpn_expr/types` already partly covers how `rpn_fn` works. Maybe you can consider consolidate them together?